core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,278 @@
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Collections2.orderedPermutations;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IrRowPatternToProgramRewriter
+        extends IrRowPatternVisitor<Integer, Void>
+{
+    public static Program rewrite(IrRowPattern node, Map<IrLabel, Integer> labelMapping)
+    {
+        ImmutableList.Builder<Instruction> builder = ImmutableList.builder();
+        new IrRowPatternToProgramRewriter(builder, labelMapping).process(node);
+        builder.add(new Done());
+        return new Program(builder.build());
+    }
+
+    private final Map<IrLabel, Integer> labelMapping;
+    private final ImmutableList.Builder<Instruction> instructions;
+    private int size;
+
+    IrRowPatternToProgramRewriter(ImmutableList.Builder<Instruction> instructions, Map<IrLabel, Integer> labelMapping)
+    {
+        this.instructions = requireNonNull(instructions, "instructions is null");
+        this.labelMapping = requireNonNull(labelMapping, "labelMapping is null");
+    }
+
+    @Override
+    protected Integer visitIrRowPattern(IrRowPattern node, Void context)
+    {
+        throw new IllegalStateException(format("unsupported node type: %s. run IrRowPatternFlattener and IrPatternAlternationOptimizer first", node.getClass().getName()));
+    }
+
+    @Override
+    protected Integer visitIrLabel(IrLabel node, Void context)
+    {
+        instructions.add(new MatchLabel(labelMapping.get(node)));
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrEmptyPattern(IrEmptyPattern node, Void context)
+    {
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrAnchorPattern(IrAnchorPattern node, Void context)
+    {
+        switch (node.getType()) {
+            case PARTITION_START:
+                instructions.add(new MatchStart());
+                break;
+            case PARTITION_END:
+                instructions.add(new MatchEnd());
+                break;
+            default:
+                throw new IllegalStateException("unexpected anchor type: " + node.getType());
+        }
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrExcludedPattern(IrExcludedPattern node, Void context)
+    {
+        instructions.add(new Save());
+        size++;
+
+        size = process(node.getPattern());
+
+        instructions.add(new Save());
+        size++;
+
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternAlternation(IrPatternAlternation node, Void context)
+    {
+        if (node.getPatterns().size() == 1) {
```

Comment:
Why would PatternAlternation ever have less than 2 entries? That seems like a bug.